daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilUnparseContentHandler.scala

```diff
@@ -0,0 +1,309 @@
+    lazy val qNameArr = qName.split(":")
+    infosetEvent.localName =
+      if (localName.nonEmpty) {
+        One(localName)
+      } else if (qNameArr.length > 1) {
+        One(qNameArr.last)
+      } else if (qNameArr.nonEmpty) {
+        One(qNameArr.head)
+      } else {
+        Nope
+      }
+    infosetEvent.namespaceURI =
+      if (uri.nonEmpty) {
+        One(uri)
+      } else if (qNameArr.length > 1) { // has a prefix
+        // this use case is for the situation where the XMLReader doesn't pass in uri/localname,
+        // so prefix information must be determined via AttributeImpl attribute. This can usually
+        // be remedied by setting the SAX_NAMESPACES_FEATURE of the XMLReader that has this class
+        // as its contentHandler.
+
+        // get the prefix off the qname
+        lazy val qNamePrefix = qNameArr.head
+        // look up prefix with and without xmlns prefix
+        if (prefixMapping != null) {
+          try {
+            One(prefixMapping.getURI(qNamePrefix))
+          } catch {
+            case _: NullPointerException => Nope
+          }
+        } else {
+          Nope
+        }
+      } else {
+        // this case is a last ditch attempt to try to get the namespace when the prefix is ""
+        // since namespacebinding doesn't take "", we convert to null.
+        try {
+          if (prefixMapping != null) {
+            One(prefixMapping.getURI(null))
+          } else {
+            Nope
+          }
+        } catch {
+          case _: NullPointerException => Nope
+        }
+      }
+    infosetEvent.eventType = One(StartElement)
+  }
+
+  override def endElement(uri: String, localName: String, qName: String): Unit = {
+    // if infosetEvent is a startElement, send that first
+    if (infosetEvent.eventType.contains(StartElement)) {
+      // any characterData that exists at this point is valid data as padding data has been
+      // taken care of in startElement
+      val maybeNewStr = One(characterData.toString)
+      infosetEvent.simpleText = maybeNewStr
+      characterData.setLength(0)
+      sendToInputter()
+    }
+
+    lazy val qNameArr = qName.split(":")
+    infosetEvent.localName =
+      if (localName.nonEmpty) {
+        One(localName)
+      } else if (qNameArr.nonEmpty) {
+        One(qNameArr.last)
+      } else {
+        Nope
+      }
+    infosetEvent.namespaceURI =
+      if (uri.nonEmpty) {
+        One(uri)
+      } else if (qNameArr.length > 1) {
+        try {
+          One(prefixMapping.getURI(qNameArr.head))
+        } catch {
+          case _: NullPointerException => Nope
+        }
+      } else {
+        try {
+          One(prefixMapping.getURI(null))
+        } catch {
+          case _: NullPointerException => Nope
+        }
+      }
+    infosetEvent.eventType = One(EndElement)
+
+    if (!contentHandlerPrefixMappingUsed) {
+      // always pops
+      prefixMapping = prefixMappingTrackingStack.pop
+    }
+    sendToInputter()
+  }
+
```

Comment:
Can the logic involved uri, localname and qname be factored out into a helper function? It looks like the logic in startElement and enedElement is the same?